### PR TITLE
DSN integration tests

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+import org.apache.james.transport.matchers.DSNFailureRequested;
+import org.apache.james.transport.matchers.DSNSuccessRequested;
+import org.apache.james.utils.DataProbeImpl;
+import org.apache.james.utils.SMTPMessageSender;
+import org.apache.james.utils.TestIMAPClient;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
```
Comment (Arsnael) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:59]: Hmm I thought we managed to reduce our junit4 tests lately to just the jmap-draft package, that we intend to deprecate and remove eventually. I know usually the answer is 'it's faster for us this way' but should we still make the effort to write new tests in JUnit5 now? what we gain now will be lost later on refactorings, that are becoming endless if we keep adding JUnit 4 tests in the code base.

WDYT?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+import org.apache.james.transport.mailets.RecipientRewriteTable;
+import org.apache.james.transport.matchers.All;
+import org.apache.james.utils.DataProbeImpl;
+import org.apache.james.utils.SMTPMessageSender;
+import org.apache.james.utils.TestIMAPClient;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
```
Comment (Arsnael) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:67]: idem

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java (hunk 1 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:59]: This module is likely a huge effort, it changes completly the estimation of the ticket IMO

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    private ProcessorConfiguration.Builder directResolutionTransport() {
+        return ProcessorConfiguration.transport()
+            .addMailet(MailetConfiguration.BCC_STRIPPER)
+            .addMailet(MailetConfiguration.builder()
+                .matcher(All.class)
+                .mailet(RecipientRewriteTable.class))
+            .addMailet(MailetConfiguration.builder()
+                .mailet(LocalDelivery.class)
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:117]: To reach the next mailet after local delivery you likely need the 'consume' option to be false (or passthrough I do not exactly remember).

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                .matcher(All.class))
+            .addMailet(MailetConfiguration.builder()
+                .matcher(DSNSuccessRequested.class)
+                .mailet(DSNBounce.class))
+            .addMailet(MailetConfiguration.builder()
+                .matcher(DSNFailureRequested.class)
+                .mailet(DSNBounce.class));
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:124]: Here we did not fail, sending failre notices is likely inapropriate

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    public void setUp() throws Exception {
+        inMemoryDNSService = new InMemoryDNSService()
+            .registerMxRecord(DEFAULT_DOMAIN, LOCALHOST_IP);
+
+        jamesServer = TemporaryJamesServer.builder()
+            .withBase(SMTP_AND_IMAP_MODULE)
+            .withOverrides(binder -> binder.bind(DNSService.class).toInstance(inMemoryDNSService))
+            .withMailetContainer(MailetContainer.builder()
+                .putProcessor(CommonProcessors.simpleRoot())
+                .putProcessor(CommonProcessors.error())
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:91]: We might want a custom error processor to send failure notices

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+    private ProcessorConfiguration.Builder directResolutionTransport() {
+        return ProcessorConfiguration.transport()
+            .addMailet(MailetConfiguration.BCC_STRIPPER)
+            .addMailet(MailetConfiguration.builder()
+                .matcher(All.class)
+                .mailet(RecipientRewriteTable.class))
+            .addMailet(MailetConfiguration.builder()
+                .mailet(LocalDelivery.class)
+                .matcher(All.class))
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:118]: You asked for  an easy way to generate failure notices and I suggested sending emails to a user of a local domain that do not exist.

To do so:

```
<mailet matcher="RecipientIsLocal" class="LocalDelivery">
     <consume>false</consume>
</mailet>
<mailet matcher="HostIsLocal" class="ToProcessor">
     <processor>error</processor>
</mailet>
<mailet matcher="All" class="ToProcessor"><processor>error</processor></mailet>
<!-- Have a DSN failure generation in your error processor -->
<mailet matcher="DSNSuccessRequested" class="DSNBounce">
    <defaultStatus>2.0.0</processor>
    <action>delivered</action>
    <prefix>[SUCCESS]</prefix>
    <messageString>Your message was successfully delivered</messageString>
</mailet>
```

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        }
+
+        testIMAPClient.connect(LOCALHOST_IP, jamesServer.getProbe(ImapGuiceProbe.class).getImapPort())
+            .login(FROM, PASSWORD)
+            .select(TestIMAPClient.INBOX)
+            .awaitNoMessage(awaitAtMostOneMinute);
+    }
+
+    @Test
+    public void givenAMailWithNotifySuccessWhenItSucceedThenADsnSuccessIsSentBack() throws IOException {
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:188]: We need stronger assert on the DSN action of the generated bounce IMO. Here it should be a relayed action?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DSNRemoteIntegrationTest {
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:71]: You do not try the behaviour when the remote server fails?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java (hunk 1 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:59]: https://github.com/linagora/james-project/pull/4221 does this :-)

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Needs a rebase obviously

**Comment by chibenwa** [comment]
test this please

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+                .mailet(DSNBounce.class)
+                .addProperty("defaultStatus", "2.0.0")
+                .addProperty("action", "delivered")
+                .addProperty("prefix", "[SUCCESS]")
+                .addProperty("messageString", "Your message was successfully delivered")
+            );
+    }
+
+    @Test
+    void givenAMailWithNoNotifyWhenItSucceedThenNoDsnIsSentBack() throws IOException {
```
[the remark this answers is not in the document]
Comment (rouazana) [reply server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNLocalIntegrationTest.java:152]: and in other places in this code

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+    private static final String FROM = "from@" + DEFAULT_DOMAIN;
+    private static final String RECIPIENT = "touser@" + ANOTHER_DOMAIN;
+    private static final ConditionFactory AWAIT_NO_MESSAGE = Awaitility.with().pollDelay(new Duration(2, TimeUnit.SECONDS)).timeout(Duration.FIVE_SECONDS);
+
+    private InMemoryDNSService inMemoryDNSService;
+    private ConfigurationClient mockSMTPConfiguration;
+
+    @RegisterExtension
+    public static DockerContainer mockSmtp = DockerContainer.fromName(MOCK_SMTP_SERVER)
+        .withLogConsumer(outputFrame -> LOGGER.debug("MockSMTP 1: " + outputFrame.getUtf8String()));
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:97]: debug?

**Review by rouazana — APPROVED** [review]
can you just cleanup the history and add the JAMES-xxxx JIRA prefix?

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java [hunk 11]
```diff
@@ -111,6 +117,17 @@ void setUp(@TempDir File temporaryFolder) throws Exception {
...
+                    .enableJmx(false)
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(IsRemoteDeliveryPermanentError.class)
+                        .mailet(ToProcessor.class)
+                        .addProperty("processor", "remote-delivery-permanent-error"))
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(IsRemoteDeliveryTemporaryError.class)
+                        .mailet(ToProcessor.class)
+                        .addProperty("processor", "remote-delivery-temporary-error"))
+                )
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:129]: Im not fan of lonely brackets

**Review by chibenwa — APPROVED** [review]

File: server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java (hunk 10 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/smtp/dsn/DSNRemoteIntegrationTest.java:97]: I don't think it hurts, it goes in the test-run.log file

**Comment by chibenwa** [comment]
> can you just cleanup the history and add the JAMES-xxxx JIRA prefix?

Can be done upon merge IMO

**Comment by chibenwa** [comment]
(Rene says: enough approuvals, hence I do not have strong need to review this, I trust you. Hence I did put the waiting merge tag.)

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**